### PR TITLE
catalog: Abstract common persist functionality

### PR DESCRIPTION
**Pull request by jkosh44**
The implementations of `UnopenedPersistCatalogState` and
`PersistCatalogState` have slowly converged to the point that they
are extremely similar and duplicate a lot of logic and
methods. This commit abstracts the common functionality into a new
struct called `PersistHandle` and implements
`UnopenedPersistCatalogState` and `PersistCatalogState` using a
`PersistHandle`.

### Motivation
This PR refactors existing code.

### Tips for reviewer

* Review the PR commit by commit. Each commit has a commit message explaining and the changes. Some commits are purely/mostly code movement and don't require in-depth review.
* The diff is smaller if viewed with whitespace hidden.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by ParkMyCar** [comment]
> @ParkMyCar do you think that the added traits will hurt the compile times for this crate?

Possibly a bit, but I wouldn't worry about it for two reasons:
1. Trying to improve the compile times with dynamic dispatch would probably add complexity because of trait objects
2. There don't appear to be _too_ many types that implement the new trait, so the compiler shouldn't have to generate too much more code than it already is.

If you're really concerned you can try `cargo build --timings` and see the output for this PR vs main

File: src/catalog/src/durable/persist.rs [hunk 1]
```diff
@@ -157,17 +158,48 @@ impl FenceableEpoch {
...
+        match self {
+            FenceableEpoch::Unfenced(Some(current_epoch)) => {
+                if epoch > *current_epoch {
+                    *self = FenceableEpoch::Fenced {
+                        current_epoch: *current_epoch,
+                        fence_epoch: epoch,
+                    };
+                    self.validate()?;
+                } else if epoch < *current_epoch {
+                    panic!("Epoch went backwards from {current_epoch:?} to {epoch:?}");
```
Comment (ParkMyCar) [inline src/catalog/src/durable/persist.rs:173]: this scares me a bit, but it already exists so I guess it's not too big of a deal?

**Review by ParkMyCar — APPROVED** [review]

File: src/catalog/src/durable/persist.rs (hunk 1 above)
Comment (jkosh44) [reply src/catalog/src/durable/persist.rs:173]: Yeah it already existed. We could change this from `panic!` to `error!`, I just worry that we'll be in some undefined state and may cause damage to the catalog if we try to continue.

**Comment by jkosh44** [comment]
@MaterializeInc/testing are the nightlies failures worrisome? They all seem to have failed recently on `main`.

**Comment by def-** [comment]
The CRDB rolling restart timeout I haven't seen before, I retried it: https://buildkite.com/materialize/nightlies/builds/7108#018e81ae-dddf-4e94-8460-abd31a266e56

**Pull request merged**